[EVA-2278] Subsequent ingestion steps

apriltuesday opened the pull request:
* create project directory
* create properties files for accession
* run accessioning pipeline
* merge vcfs if needed
* create properties files for variant load
* run variant load pipeline

tcezard commented on eva_submission/ingestion_templates.py:
```diff
@@ -0,0 +1,119 @@
+
+output.dir=/nfs/production3/eva/data/{project_accession}/40_transformed/
+output.dir.annotation=/nfs/production3/eva/data/{project_accession}/51_annotation/
+output.dir.statistics=/nfs/production3/eva/data/{project_accession}/50_stats/
```
We should avoid having path from the cluster in public github. Get the path from the config file.

apriltuesday replied:
thank you for catching, these should all be removed

tcezard commented on eva_submission/ingestion_templates.py:
```diff
@@ -0,0 +1,119 @@
+
+output.dir=/nfs/production3/eva/data/{project_accession}/40_transformed/
+output.dir.annotation=/nfs/production3/eva/data/{project_accession}/51_annotation/
+output.dir.statistics=/nfs/production3/eva/data/{project_accession}/50_stats/
```
We should avoid having path from the cluster in public github. Get the path from the config file.

tcezard commented on eva_submission/eload_ingestion.py:
```diff
@@ -16,33 +37,71 @@ class EloadIngestion(Eload):
+            tasks=None
+    ):
         # TODO assembly/taxonomy insertion script should be incorporated here
+        # TODO set ENA data release date
```
```suggestion
        # TODO set ENA data release date (EVA-2322)
```

tcezard commented on bin/ingest_submission.py:
```diff
@@ -31,6 +31,11 @@
 def main():
     argparse = ArgumentParser(description='Accession and ingest submission data into EVA')
     argparse.add_argument('--eload', required=True, type=int, help='The ELOAD number for this submission')
+    argparse.add_argument('--instance', required=False, type=int, help='The instance id to use for accessioning')
```
```suggestion
    argparse.add_argument('--instance', required=False, type=int, choices=range(1, 13), help='The instance id to use for accessioning')
```

tcezard commented on eva_submission/eload_ingestion.py:
```diff
@@ -16,33 +37,71 @@ class EloadIngestion(Eload):
+            self.eload_cfg.set(self.config_section, 'aggregation', value=aggregation)
+
+        if do_accession:
+            if instance_id not in range(1, 13):
+                raise ValueError('Instance id must be between 1-12')
```
This can be checked in the command line interface

tcezard commented on eva_submission/eload_ingestion.py:
```diff
@@ -16,33 +37,71 @@ class EloadIngestion(Eload):
+        if do_accession or do_variant_load:
+            aggregation = aggregation.lower()
+            if aggregation not in {'basic', 'none'}:
+                raise ValueError('Aggregation type must be BASIC or NONE')
```
Is this required since it's already tested in the CLI?

apriltuesday replied:
indeed, removed

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,106 @@
+        path "accession.properties" from accession_props
+
+    output:
+        path "00_logs/accessioning.*.log" into accessioning_log
+        path "00_logs/accessioning.*.err" into accessioning_err
+        path "60_eva_public/*.vcf" into accessioned_vcfs
```
This comment applies to many steps in the workflows:
In Nextflow, the script is being executed in a temp directory placed in the current directory which usually looks like this: `work/91/eff364f4a21f7a5df088022564c7bd`. 
The output that you declare above will be search in this temp directory and when not found, the step will fail.
I think the best way would be to get the `00_logs` and `60_eva_public` as input so they are linked in the temp directory, then you can have the files as an output. 
This said it mean you can't use the wildcards because you can't guarantee that there won't be any other VCF/logs in those folders.

apriltuesday replied:
This and the other input/output issues should be fixed, please have a look and let me know.

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,106 @@
+ */
+process compress_vcfs {
+    input:
+        path vcf_file from accessioned_vcfs
```
Just to add to what I've said above, If you get this accessioned vcf as an input, the `60_eva_public` won't exist in the folder.

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,106 @@
+    $params.executable.tabix -p vcf $compressed_vcf
+    """
+}
+
```
I would suggest also adding the bcftools index (I know it's not a standard step at the moment) because it will be required in the future. See [here](https://github.com/EBIvariation/eva-submission/blob/d03305880a6b215ced29e3aecc6a2fa1207ba3d3/nextflow/prepare_brokering.nf#L67) for an example

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,106 @@
+
+    """
+    cd 60_eva_public
+    $params.executable.copy_to_ftp $params.project_accession
```
Is this a place holder because I can't find any `copy_to_ftp ` in the executable

tcezard reviewed (changes requested):
This is a big piece of work, well done. 
Happy to help with the nextflow input/output management as it is a bit tricky.

apriltuesday commented:
@tcezard @sundarvenkata-EBI @andresfsilva This should be ready for review when you have some time.

Just to note, if you started taking a look at the Python last week, that code is almost unchanged, but I have updated the Nextflow substantially and added test scripts.  Those changes are encapsulated in [this commit](https://github.com/EBIvariation/eva-submission/pull/17/commits/15c58db4723559bdad8ebe640ed24b104e57bd38).

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -16,66 +37,84 @@ class EloadIngestion(Eload):
         self.settings_xml_file = cfg['maven_settings_file']
+        self.project_accession = self.eload_cfg.query('brokering', 'ena', 'PROJECT')
+        self.project_dir = self.setup_project_dir()
+        self.mongo_uri = get_mongo_uri_for_eva_profile('production', self.settings_xml_file)
```
Judgment call but I am not sure hard-coding production environment is advisable here... Sometimes it may be worthwhile to test the automation in an [internal environment](https://github.com/EBIvariation/configuration/blob/22cb2ac3e675ae1913849fcaefc51bed2e2f69e1/eva-maven-settings.xml#L27) just to make sure everything works properly.

apriltuesday replied:
Good point, I'll make it a configurable value.

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -16,66 +37,84 @@ class EloadIngestion(Eload):
+            self.eload_cfg.set(self.config_section, 'variant_load', 'vep', 'cache_version', value=vep_cache_version)
+            if aggregation == 'none':
+                merged_files = self.merge_vcfs()
+                self.vcfs_to_load = merged_files
```
```suggestion
                self.vcfs_to_load = self.merge_vcfs()
```

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+            return [Path(f'{output_file}.gz')]
+        except subprocess.CalledProcessError as e:
+            if e.output is not None and 'Duplicate sample names' in e.output:  # TODO brittle
+                self.warning('Duplicate sample names found while merging, will continue with unmerged VCFs.')
```
I am not sure if it is prudent to do this... Let's discuss further in DEV meeting...

apriltuesday replied:
Captured in [EVA-2334](https://www.ebi.ac.uk/panda/jira/browse/EVA-2334), will update the `TODO` to refer to the ticket

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+            filename = vcf_path.stem
+            output_vcf = self.project_dir.joinpath(project_dirs['public'], f'{filename}.accessioned.vcf')
+
+            properties_filename = self.project_dir.joinpath(project_dirs['accessions'], f'{filename}.properties')
```
```suggestion
            properties_filename = self.project_dir.joinpath(project_dirs['accessions'], f'{filename}_accessioning.properties')
```

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+
+    def get_mongo_creds(self):
+        properties = get_properties_from_xml_file('production', self.settings_xml_file)
+        mongo_host = str(str(properties['eva.mongo.host']).split(',')[0]).split(':')[0]
```
Please use the official [Mongo URI parser](https://pymongo.readthedocs.io/en/stable/api/pymongo/uri_parser.html) to keep up with future changes (ex: MongoDB 4.0 cluster will possibly get authMechanism parameter).

apriltuesday replied:
I'll definitely do this, but this reminds me of [another question](https://github.com/apriltuesday/eva-submission/blob/EVA-2278/eva_submission/eload_ingestion.py#L220) that I had - basically the only reason I need this method rather than just `get_mongo_uri_for_eva_profile` is that the accessioning pipeline needs the credentials in the properties file.  Is there a way the pipeline can get those for itself in the future?

sundarvenkata-EBI replied:
The pipelines were created with the goal of being installable in other environments besides EVA. Therefore, it might be tricky. But it is something that we can look into. Could you create a ticket for that in JIRA?

apriltuesday replied:
Right, that makes sense.  I'm imagining something like what eva-pipeline does, with a production properties file containing the credentials and a secondary one created per submission.  But of course I don't know if that's possible for accessioning, and it's definitely a nice-to-have from my perspective.

apriltuesday replied:
Either way I've created [EVA-2336](https://www.ebi.ac.uk/panda/jira/browse/EVA-2336) so we can discuss/investigate.

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+    def get_pg_conn(self):
+        pg_url, pg_user, pg_pass = self.get_pg_creds()
+        # need to cut the jdbc: from the front of the pg_url
+        return psycopg2.connect(pg_url[5:], user=pg_user, password=pg_pass)
```
Please investigate the urllib.parse [urlsplit](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlsplit) to do this in a standardized way.

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+
+    def get_vep_species(self):
+        words = self.eload_cfg.query('submission', 'scientific_name').lower().split()
+        return '_'.join(words)
```
Sometimes Ensembl scientific names are messed up. May be you would want to use a logic similar to [this](https://github.com/EBIvariation/eva-common-pyutils/blob/702ccb71be5479e66847bb7424c53597a5fc65de/ebi_eva_common_pyutils/taxonomy/taxonomy.py#L32) and see if it matches with some of the VEP caches that are currently available.

apriltuesday replied:
As posted in Slack, the existing cache names match this naming scheme so this is probably okay for now, but as we don't currently control the naming we can update this to something more robust as needed.

sundarvenkata-EBI commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+                ))
+            )
+        except subprocess.CalledProcessError as e:
+            self.error('Nextflow accessioning pipeline failed: results might not be complete')
```
Could  the Nextflow log file be added to the error message for debugging purposes?

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,133 @@
+ */
+process accession_vcf {
+    clusterOptions '-g /accession/instance-$params.instance_id'
+
```
I would suggest adding a memory requirement for the cluster
```suggestion

    memory 8 GB
    
```

tcezard commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,133 @@
+accession_props = Channel.fromPath(params.accession_props)
+num_props = Channel.fromPath(params.accession_props).count().value
+// watches public dir for the same number of vcf files as there are property files
+accessioned_vcfs = Channel.watchPath(params.public_dir + '/*.vcf').take(num_props)
```
That's an interesting technique I did not know existed. 
I'm still a bit worried about other VCF laying around the public directory but we can solve this another way by ensuring that we only process one submission per directory.

apriltuesday replied:
Just to confirm the behaviour of this, if there are VCFs in the public directory before the pipeline starts, `watchPath` will ignore them (including if they're overwritten).  But if someone adds some random VCFs to the directory while the accessioning is running, this method won't be able to tell the difference.  I'll add a comment about this and we'll revisit in the future.

apriltuesday replied:
Actually, I wonder if I could filter the channel to only include VCFs that explicitly match the properties files?  Might be a little safer even if we still have to revisit the larger issue.

tcezard commented on nextflow/variant_load.nf:
```diff
@@ -0,0 +1,49 @@
+* Load into variant db.
+*/
+process load_vcf {
+    input:
```
```suggestion
    memory 5 GB

    input:
```

tcezard reviewed (approved):
Excellent work. most of the comments are minor and the ones aren't should be address in other tickets

sundarvenkata-EBI commented on nextflow/accession.nf:
```diff
@@ -0,0 +1,133 @@
+params.help = null
+
+// Show help message
+if (params.help) exit 0, helpMessage()
```
So even if we explicitly set params.help to null in the previous line, this will still work because we pass in the command line parameter?

apriltuesday replied:
That's my understanding, and I just tried it to check that it works...  To be honest I find this a bit unintuitive as well, but I modelled it off @tcezard 's code [here](https://github.com/EBIvariation/eva-submission/blob/master/nextflow/prepare_brokering.nf#L22).

tcezard replied:
Actually this is fine as it is.
`params.help` is initialise as `null` and then can be overridden by the command line. 
If it is not set on the command line then it is still `null` and interpreted as `false` otherwise it is set to `true` and show the help message.
The only weird bit is that we don't see where the command line args are being parsed and assigned to the params object. but that's the :mage_woman: magic of a DSL :mage:

apriltuesday replied:
Makes sense, thanks for explaining!

sundarvenkata-EBI reviewed (approved):
Major league stuff here! I enjoyed reviewing it and also learnt a lot from it! Left a few comments but we can address those in the next iteration.

apriltuesday commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+        with open(list_file, 'w+') as f:
+            f.write('\n'.join(str(fn) for fn in self.valid_vcf_filenames))
+        try:
+            command_utils.run_command_with_output(
+                'Merge VCFs',
+                ' '.join((
+                    cfg['executable']['bcftools'], 'merge'
+                    '--merge', 'all',
+                    '--file-list', list_file,
+                    '--threads', '3',
+                    '-o', output_file
+                ))
+            )
+            command_utils.run_command_with_output(
+                'Bgzip merged vcf file',
+                ' '.join((cfg['executable']['bgzip'], output_file))
+            )
```
Captured in [EVA-2331](https://www.ebi.ac.uk/panda/jira/browse/EVA-2331)

apriltuesday commented on eva_submission/eload_ingestion.py:
```diff
@@ -114,3 +152,215 @@ def load_from_ena(self):
+        # copy valid vcfs + index to 'valid' folder
+        valid_dir = project_dir.joinpath(project_dirs['valid'])
+        vcf_dict = self.eload_cfg.query('brokering', 'vcf_files')
+        if vcf_dict is None:
+            self.error('No brokered VCF files found, aborting ingestion.')
+            raise ValueError('No brokered VCF files found.')
```
Good point, I've collected the existing checks in a single function run at the beginning of the `ingest` function, will make it easier to add any validations if they become necessary
